softioc/autosave.py

```diff
@@ -0,0 +1,178 @@
+
+    def _change_directory(self, directory: str):
+        dir_path = Path(directory)
+        if dir_path.is_dir():
+            self.directory = dir_path
+        else:
+            raise ValueError(f"{directory} is not a valid autosave directory")
+
+    def backup_sav_file(self):
+        sav_path = self._get_current_sav_path()
+        if sav_path.is_file():
+            shutil.copy2(sav_path, self._get_timestamped_backup_sav_path())
+
+    @classmethod
+    def add_pv(cls, pv):
+        pv.set_autosave(True)
+        cls._pvs[pv._name] = pv
+
+    @classmethod
+    def remove_pv(cls, pv):
+        pv.set_autosave(False)
+        cls._pvs.pop(pv._name, None)
+
+    def _get_timestamped_backup_sav_path(self):
+        sav_path = self._get_current_sav_path()
+        return sav_path.parent / (
+            sav_path.name + self._last_saved_time.strftime("_%y%m%d-%H%M%S")
+        )
+
+    def _get_backup_save_path(self):
+        return self.directory / f"{self.device_name}.{SAVB_SUFFIX}"
+
+    def _get_current_sav_path(self):
+        return self.directory / f"{self.device_name}.{SAV_SUFFIX}"
+
+    def _save(self, state):
+        try:
+            for path in [
+                self._get_current_sav_path(),
+                self._get_backup_save_path()
+            ]:
+                with open(path, "w") as f:
+                    yaml.dump(state, f, indent=4)
+            self._last_saved_state = state.copy()  # do we need to copy?
+            self._last_saved_time = datetime.now()
+        except Exception as e:
+            print(f"Could not save state to file: {e}")
+
+    def save(self):
+        if not self.enabled:
+            print("Not saving to file as autosave adapter disabled")
+            return
+        state = {name: pv.get() for name, pv in self._pvs.items()}
+        if state != self._last_saved_state:
+            self._save(state)
+
+    def load(self, path=None):
+        if not self.enabled:
+            print("Not loading from file as autosave adapter disabled")
+            return
+        sav_path = path or self._get_current_sav_path()
+        if not sav_path or not sav_path.is_file():
+            print(f"Could not load autosave values from file {sav_path}")
+            return
+        with open(sav_path, "r") as f:
+            state = yaml.full_load(f)
+        for name, value in state.items():
+            pv = self._pvs.get(name, None)
+            if not pv:
+                print(f"{name} is not a valid autosaved PV")
+                continue
+            pv.set(value)
+
+    def stop(self):
+        self._stop_event.set()
+
+    def loop(self):
+        if not self._pvs:
+            return  # end thread if no PVs to save
+        while True:
+            self._stop_event.wait(timeout=self.save_period)
+            if self._stop_event.is_set():  # Stop requested
+                return
+            else:  # No stop requested, we should save and continue
+                self.save()
```

Comment:
We need to consider how this should behave for long-running applications. Currently if there is any exception the entire thread will end up terminating. We probably don't want that, especially as it's possible that the exception could be from a temporary or one-time issue (e.g. network file mount blip).

We probably want at least a `try-except` clause here at this top level.